src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentAction.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Options;
+
+namespace Microsoft.CodeAnalysis.Rename
+{
+    public static partial class Renamer
+    {
+        /// <summary>
+        /// Individual action from RenameDocument APIs in <see cref="Renamer"/>.
+        /// 
+        /// See <see cref="RenameDocumentActionSet" />
+        /// </summary>
+        public abstract class RenameDocumentAction
+        {
+            private readonly ImmutableArray<ErrorResource> _errorStringKeys;
+            protected OptionSet OptionSet { get; }
+
+            public virtual ImmutableArray<string> GetErrors(CultureInfo? culture = null)
+                => _errorStringKeys.SelectAsArray(s => string.Format(WorkspacesResources.ResourceManager.GetString(s.FormatString, culture ?? WorkspacesResources.Culture)!, s.Arguments));
+
+            public abstract string GetDescription(CultureInfo? culture = null);
+            internal abstract Task<Solution> GetModifiedSolutionAsync(Document document, CancellationToken cancellationToken);
+
+            internal RenameDocumentAction(ImmutableArray<ErrorResource> errors, OptionSet optionSet)
+            {
+                OptionSet = optionSet;
+                _errorStringKeys = errors;
+            }
+
+            internal readonly struct ErrorResource
+            {
+                public string FormatString { get; }
+                public object?[] Arguments { get; }
```

Comment:
when is an argument null?

Reply:
whenever the string doesn't need to be formatted. It's valid for `string.Format` calls

Reply:
wouldn'thtat be `object[]?` not `object?[]`?  the latter looks like an array of nullable objects.

Reply:
I followed the annotation for `string.Format` on this one. I must admit I find the null annotation on arrays rather confusing...

Reply:
Indeed this is an array of nullable objects, but the array itself can't be null. Doesn't quite sound like that's what is being described here...?

Reply:
Removing `?`